thirdparty/faiss/faiss/cppcontrib/knowhere/impl/HnswSearcher.h

```diff
@@ -0,0 +1,415 @@
+// Copyright (C) 2019-2024 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#pragma once
+
+// standard headers
+#include <algorithm>
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+
+// Faiss-specific headers
+#include <faiss/Index.h>
+#include <faiss/IndexHNSW.h>
+#include <faiss/impl/AuxIndexStructures.h>
+#include <faiss/impl/DistanceComputer.h>
+#include <faiss/impl/FaissException.h>
+#include <faiss/impl/HNSW.h>
+#include <faiss/utils/ordered_key_value.h>
+
+// Knowhere-specific headers
+#include <faiss/cppcontrib/knowhere/impl/Neighbor.h>
+
+
+namespace faiss {
+namespace cppcontrib {
+namespace knowhere {
+
+namespace {
+
+// whether to track statistics
+constexpr bool track_hnsw_stats = true;
+
+}
+
+// Accomodates all the search logic and variables.
+/// * DistanceComputerT is responsible for computing distances
+/// * GraphVisitorT records visited edges
+/// * VisitedT is responsible for tracking visited nodes
+/// * FilterT is resposible for filtering unneeded nodes
+/// Interfaces of all templates are tweaked to accept standard Faiss structures
+///   with dynamic dispatching. Custom Knowhere structures are also accepted.
+template<typename DistanceComputerT, typename GraphVisitorT, typename VisitedT, typename FilterT>
+struct v2_hnsw_searcher {
```

Comment:
Plz correct me if I am wrong, this is the same logic as what Knowhere' hnsw is right now

Reply:
Originally, hnswlib contained a code for a brute-force search. So, I kinda reimplemented it. My goal was to make sure that the QPS value matches.

Reusing what's available in Faiss needs to be evaluated.

Reply:
technically, I'm not putting things in Faiss. My future goal is to use a vanilla Faiss and have a dedicated set of additional components that belong to knowhere. I'd like to keep these Knowhere components in `faiss/cppcontrib/knowhere` directory.

Reply:
> technically, I'm not putting things in Faiss. My future goal is to use a vanilla Faiss and have a dedicated set of additional components that belong to knowhere. I'd like to keep these Knowhere components in `faiss/cppcontrib/knowhere` directory.

OK, it means that this implementation is from Knowhere's HNSW. To my understanding, this implementation is some kind of temporary fall back plan and need to be removed after the main implementation is stable. So let it be where it used to be (knowhere/thirdparty/hnswlib) can make the implementation easier? 

Just curios about any concerns here, I don't have a strong opinion to put it anywhere.

Reply:
We can always change locations later, this is not important at this moment.

Reply:
More than location. 
1. Since we have three HNSW for now: The new implemented one, /knohwere/thirdpart/hnswlib, and this searcher.
2. Given the purpose is fallback to old solution, old code can always provide more promising old solution

It is a little bit hard for keep them updated (We may need to do so for temporary).

I will create a issue to follow and unblock this PR. We can have a further discussion there.

Reply:
https://github.com/zilliztech/knowhere/issues/776